### PR TITLE
lightning: use gRPC prepared message to avoid duplicated encoding

### LINKED ISSUE #45383: lightning use grpc PreparedMsg to avoid duplicate encoding
## Enhancement

https://pkg.go.dev/google.golang.org/grpc#PreparedMsg.Encode

lightning will send same `pairs` to every peers 
https://github.com/pingcap/tidb/blob/5027679c6a167828961ea3fa03f47f7f4619442d/br/pkg/lightning/backend/local/region_job.go#L273-L279

if we can convert WriteRequest_Batch to a PreparedMsg, we can avoid repeatedly encoding in grpc framework

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: close #45383

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

before
```
[2023/07/24 14:01:26.934 +08:00] [INFO] [local.go:1462] ["import engine success"] [uuid=faa8289d-3378-544f-8e7a-c12498f388d4] [size=108087902895] [kvs=100000000] [importedSize=108087902895] [importedCount=100000000]
[2023/07/24 14:01:26.935 +08:00] [INFO] [backend.go:353] ["import completed"] [engineTag=`test`.`usertable`:0] [engineUUID=faa8289d-3378-544f-8e7a-c12498f388d4] [retryCnt=0] [takeTime=3m41.016948476s] []
[2023/07/24 14:01:52.803 +08:00] [INFO] [local.go:1462] ["import engine success"] [uuid=bede5e32-334c-5343-8ef3-08a54938c5c3] [size=5499999973] [kvs=100000000] [importedSize=5499999973] [importedCount=100000000]
[2023/07/24 14:01:52.804 +08:00] [INFO] [backend.go:353] ["import completed"] [engineTag=`test`.`usertable`:-1] [engineUUID=bede5e32-334c-5343-8ef3-08a54938c5c3] [retryCnt=0] [takeTime=23.40315731s] []
[2023/07/24 14:01:53.497 +08:00] [INFO] [import.go:515] ["the whole procedure completed"] [takeTime=14m38.509928185s] []
```

after
```
[2023/07/24 14:49:10.305 +08:00] [INFO] [local.go:1462] ["import engine success"] [uuid=faa8289d-3378-544f-8e7a-c12498f388d4] [size=108087902895] [kvs=100000000] [importedSize=108087902895] [importedCount=100000000]
[2023/07/24 14:49:10.306 +08:00] [INFO] [backend.go:353] ["import completed"] [engineTag=`test`.`usertable`:0] [engineUUID=faa8289d-3378-544f-8e7a-c12498f388d4] [retryCnt=0] [takeTime=2m54.358068332s] []
[2023/07/24 14:49:31.263 +08:00] [INFO] [local.go:1462] ["import engine success"] [uuid=bede5e32-334c-5343-8ef3-08a54938c5c3] [size=5499999973] [kvs=100000000] [importedSize=5499999973] [importedCount=100000000]
[2023/07/24 14:49:31.264 +08:00] [INFO] [backend.go:353] ["import completed"] [engineTag=`test`.`usertable`:-1] [engineUUID=bede5e32-334c-5343-8ef3-08a54938c5c3] [retryCnt=0] [takeTime=18.383461449s] []
[2023/07/24 14:49:31.964 +08:00] [INFO] [import.go:511] ["the whole procedure completed"] [takeTime=13m48.139963409s] []
```

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/lightning/backend/local/region_job.go [hunk 1]
```diff
@@ -271,12 +270,19 @@ func (local *Backend) writeToTiKV(ctx context.Context, j *regionJob) error {
...
+			return err
+		}
+
 		for i := range clients {
 			if err := writeLimiter.WaitN(ctx, allPeers[i].StoreId, int(size)); err != nil {
 				return errors.Trace(err)
 			}
-			requests[i].Chunk.(*sst.WriteRequest_Batch).Batch.Pairs = pairs[:count]
-			if err := clients[i].Send(requests[i]); err != nil {
+			if err := clients[i].SendMsg(preparedMsg); err != nil {
```
Comment (D3Hunter) [inline br/pkg/lightning/backend/local/region_job.go:285]: seems we need use RecvMsg to ensure delivery
```
	// SendMsg does not wait until the message is received by the server. An
	// untimely stream closure may result in lost messages. To ensure delivery,
	// users should ensure the RPC completed successfully using RecvMsg.
```

**Review by D3Hunter — APPROVED** [review]
lgtm, except that prepared msg is an experimental feature

**Review by GMHDBJD — APPROVED** [review]
LGTM

**Pull request merged**